src/lib.rs

```diff
@@ -299,92 +322,206 @@ pub fn write_b(cw: &mut ContextWriter, fi: &FrameInvariants, fs: &mut FrameState
+                // It works now because we set mode for each tx position, where tx_size is 4x4.
+                cw.bc.set_mode(&tx_bo, mode);
+                write_tx_b(fi, fs, cw, p, &tx_bo, mode, tx_type, &po);
             }
         }
     }
-    let uv_tx_type = exported_intra_mode_to_tx_type_context[uv_mode as usize];
-    for p in 1..3 {
-        for by in 0..8 {
-            for bx in 0..8 {
-                let bo = sbo.block_offset(bx, by);
-                write_b(cw, fi, fs, p, &bo, uv_mode, uv_tx_type);
+
+    if has_chroma(&bo, bsize, xdec, ydec) {
+        let uv_tx_type = exported_intra_mode_to_tx_type_context[uv_mode as usize];
+        let partition_x = (bo.x & LOCAL_BLOCK_MASK) >> xdec << MI_SIZE_LOG2;
+        let partition_y = (bo.y & LOCAL_BLOCK_MASK) >> ydec << MI_SIZE_LOG2;
+
+        for p in 1..3 {
+            let sb_offset = bo.sb_offset().plane_offset(&fs.input.planes[p].cfg);
+
+            for by in 0..bh >> ydec {
+                for bx in 0..bw >> xdec {
+                    let tx_bo = BlockOffset{x: bo.x + bx, y: bo.y + by};
+                    let po = PlaneOffset {
+                        x: sb_offset.x + partition_x + (bx << MI_SIZE_LOG2),
+                        y: sb_offset.y + partition_y + (by << MI_SIZE_LOG2) };
+
+                    write_tx_b(fi, fs, cw, p, &tx_bo, uv_mode, uv_tx_type, &po);
+                }
             }
         }
     }
+}
+
+// RDO-based mode decision
+fn rdo_mode_decision(fi: &FrameInvariants, fs: &mut FrameState,
+                  cw: &mut ContextWriter,
+                  bsize: BlockSize, bo: &BlockOffset) -> RDOOutput {
+    let q = dc_q(fi.qindex) as f64;
+    let q0 = q / 8.0_f64;	// Convert q into Q0 precision, given thatn libaom quantizers are Q3.
+
+    // Lambda formula from doc/theoretical_results.lyx in the daala repo
+    let lambda = q0*q0*2.0_f64.log2()/6.0;	// Use Q0 quantizer since lambda will be applied to Q0 pixel domain
+
+    let mut best_mode = PredictionMode::DC_PRED;
+    let mut best_rd = std::f64::MAX;
+    let tell = cw.w.tell_frac();
+    let w = block_size_wide[bsize as usize];
+    let h = block_size_high[bsize as usize];
+
+    for &mode in RAV1E_INTRA_MODES {
+        let checkpoint = cw.checkpoint();
+
+        write_b(fi, fs, cw, mode, bsize, bo);
+        let po = bo.plane_offset(&fs.input.planes[0].cfg);
+        let d = sse_wxh(&fs.input.planes[0].slice(&po), &fs.rec.planes[0].slice(&po),
+                        w as usize, h as usize);
+        let r = ((cw.w.tell_frac() - tell) as f64)/8.0;
+
+        let rd = (d as f64) + lambda*r;
+        if rd < best_rd {
+            best_rd = rd;
+            best_mode = mode;
+        }
+
+        cw.rollback(checkpoint.clone());
+    }
+
+    assert!(best_rd as i64 >= 0);
+
+    let rdo_output = RDOOutput { rd_cost: best_rd as u64,
+                                pred_mode: best_mode};
+    rdo_output
+}
+
+fn write_sb(fi: &FrameInvariants, fs: &mut FrameState, cw: &mut ContextWriter,
```

Comment:
This function needs to be renamed. It's really writing a partition, not a superblock (sb). Maybe write_partition?

Reply:
We have been already using write_partition() to encode partition symbols to bitstream buffer.
Meanwhile, I think encode_partition() or write_partition() both can have dual meanings, one for partition symbols only, another for all the info in a partition, i.e. tx splits, coeffs, everything coded to decode the pixels in the partition.
Maybe 'partition_block' can better mean the pixels in the partition. So, encode_partition_block() or write_partition_block()?
One thing is for sure. Calling encode_sb() recursively as partition keep split would be avoided.

Most preferred by me is to sync with AV1 syntax spec, which uses 'decode_partition()' and 'decode_block()', so following them would at least help people dependent on AV1 spec.

So, I am trying below changes regarding function names, that I believe matches AV1 spec:
write_sb() --> encode_partition()
write_b()  --> encode_block()
write_tx_b()  --> encode_tx_block()  // Or, encode_tx_residual?

Please let me know how this look like.